router_types.c

```diff
@@ -44,6 +49,53 @@ static const struct router_type *router_types[] = {
 	NULL,
 };
 
+static int read_mac(uint8_t mac[ETH_ALEN], const char *macstr)
+{
+	int ret;
+
+	ret = sscanf(macstr, "%02hhX:%02hhX:%02hhX:%02hhX:%02hhX:%02hhX",
+		     &mac[0], &mac[1], &mac[2], &mac[3], &mac[4], &mac[5]);
+	if (ret != 6)
+		ret = sscanf(macstr, "%02hhX-%02hhX-%02hhX-%02hhX-%02hhX-%02hhX",
+			     &mac[0], &mac[1], &mac[2], &mac[3], &mac[4], &mac[5]);
+
+	return (ret == 6);
+}
+
+int mac_whitelist_add(const char *macstr)
+{
+	uint8_t mac[ETH_ALEN];
+
+	if (!read_mac(mac, macstr)) {
+		fprintf(stderr, "Error - could not add MAC address to whitelist: %s\n", macstr);
+		return -1;
+	}
+
+	struct mac_whitelist_entry *new_entry = malloc(sizeof(*mac_whitelist_head));
```

Comment:
this must now (check the sizeof parameter) be 

    new_entry = malloc(sizeof(*new_entry));

And please variable definitions to the beginning of the function